### PR TITLE
Add Unit tests for VForm, VField and VBlock

### PR DESCRIPTION
**Pull request by MedAliMarz**
This PR adds Unit tests for functionalities inside 3 classes:
- VForm
- VField
- VBlock

### CODE REVIEW

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt [hunk 1]
```diff
@@ -377,4 +381,54 @@ class VFieldTests : JApplicationTestBase() {
...
+    assertEquals(-1, field.fetchColumn(User))
+  }
+
+  @Test
+  fun `enter test valid scenario` () {
+    val field = FormSample.tb1.blockFields[3].vField
+
+    field.block?.enter()
+    field.block?.activeField?.leave(false)
+    field.enter()
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt:405]: the test work these two lines

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt [hunk 2]
```diff
@@ -377,4 +381,54 @@ class VFieldTests : JApplicationTestBase() {
...
+    assertFails { field.enter() }
+  }
+
+  @Test
+  fun `leave test valid scenario` () {
+    val field = FormSample.tb1.blockFields[3].vField // 3rd field in the first visible field
+
+    field.block?.enter()
+    field.leave(false)
+    assertNotEquals(field, field.block?.activeField)
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt:424]: you can also add : assertNull(field.block?.activeField)

**Review by h-haddad — CHANGES_REQUESTED** [review]

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/FormSampleTests.kt [hunk 3]
```diff
@@ -41,4 +71,161 @@ class FormSampleTests: JApplicationTestBase() {
...
+  @Test
+  fun `setCommandsEnabled test`() {
+    val form = FormSample()
+    val formModel = form.model
+
+    formModel.prepareForm() // internally will invoke setCommandsEnabled(true)
+    assertTrue(formModel.blocks.any {
+      it.getActiveCommands().isNotEmpty()
+    })
+  }
```
[the remark this answers is not in the document]
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/form/FormSampleTests.kt:106]: Changed it to use directly the `setCommandsEnabled` without calling prepareForm, that way we test the isolated functionality

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/FormSampleTests.kt [hunk 4]
```diff
@@ -41,4 +71,161 @@ class FormSampleTests: JApplicationTestBase() {
...
+
+  @Test
+  fun `gotoPage test`() {
+    val form = FormSample()
+    val formModel = form.model
+    val targetPage = 0
+
+    formModel.gotoPage(targetPage)
+    assertEquals(form.formBlocks[0].vBlock, formModel.getActiveBlock())
+  }
```
[the remark this answers is not in the document]
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/form/FormSampleTests.kt:116]: Done

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt [hunk 5]
```diff
@@ -932,4 +935,163 @@ class VBlockTests : VApplicationTestBase() {
...
+  @Test
+  fun `setColor test`(){
+    formMultiple.model
+    val model = formMultiple.multipleBlock.vBlock
+    val recordId = model.recordCount
+
+    model.setColor(recordId, VColor.BLACK, VColor.WHITE)
+    assertTrue(model.fields.filter { !it.isInternal() }.all{
+      it.getForeground(recordId) == VColor.BLACK && it.getBackground(recordId) == VColor.WHITE
+    })
```
[the remark this answers is not in the document]
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt:974]: Every record is linked to multiple fields, and the setColor change them all, therefore it is needed to verify that all of these fields' color has changed. Check the implementation of the `.setColor` method.

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt [hunk 6]
```diff
@@ -932,4 +935,163 @@ class VBlockTests : VApplicationTestBase() {
...
+  @Test
+  fun `resetColor test`(){
+    formMultiple.model
+    val model = formMultiple.multipleBlock.vBlock
+    val recordId = model.recordCount
+
+    model.resetColor(recordId)
+    assertTrue(model.fields.filter { !it.isInternal() }.all{
+      it.getForeground(recordId) == null && it.getBackground(recordId) == null
+    })
```
[the remark this answers is not in the document]
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt:986]: Same as above

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt [hunk 7]
```diff
@@ -932,4 +935,163 @@ class VBlockTests : VApplicationTestBase() {
...
+    assertEquals(formMultiple.saveBlock.model, model.getActiveCommands().get(0)?.actor)
+  }
+
+  @Test
+  fun `setMode test`(){
+    formMultiple.model
+    val model = formMultiple.multipleBlock.vBlock
+
+    model.setMode(VConstants.MOD_INSERT)
+    assertEquals(model.getMode(), VConstants.MOD_INSERT)
```
[the remark this answers is not in the document]
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt:955]: Done

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt [hunk 8]
```diff
@@ -932,4 +935,163 @@ class VBlockTests : VApplicationTestBase() {
...
+  fun `leave test`(){
+    val formSample = FormSample().also { it.model }
+    val blockModel = formSample.tb1.vBlock
+
+    transaction {
+      initSampleFormTables()
+      blockModel.enter()
+      formSample.model.getActiveBlock()?.leave(false)
+      assertNotEquals(formSample.model.getActiveBlock(), blockModel)
+      assertNotEquals(formSample.tb1.blockFields[3].vField, blockModel.activeField)
```
[the remark this answers is not in the document]
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt:999]: I verified that the first visible field (the 4th field in the block, indexed 3) isn't active after leaving the block. as `.leave` method should nullify the `activeField`

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt [hunk 9]
```diff
@@ -932,4 +935,163 @@ class VBlockTests : VApplicationTestBase() {
...
+
+  @Test
+  fun `getNumberOfValidRecord test`(){
+    val formSample = FormSample().also { it.model }
+    val blockModel = formSample.tb1.vBlock
+
+    transaction {
+      initSampleFormTables()
+      assertEquals(1, blockModel.numberOfValidRecord)
+      assertEquals(1, blockModel.numberOfFilledRecords)
```
[the remark this answers is not in the document]
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt:1031]: Yes, that's more suitable

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt [hunk 10]
```diff
@@ -932,4 +935,163 @@ class VBlockTests : VApplicationTestBase() {
...
+    val singleBlockModel = FormWithList.block3.vBlock
+
+    transaction {
+      FormWithList.block3.load()
+
+      singleBlockModel.trailRecord(0)
+      assertTrue(singleBlockModel.isRecordTrailed(0))
+
+      singleBlockModel.trailRecord(singleBlockModel.currentRecord)
+      assertTrue(singleBlockModel.isRecordTrailed(singleBlockModel.currentRecord))
```
[the remark this answers is not in the document]
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt:1074]: Removed redundancy

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt (hunk 1 above)
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt:405]: The `.enter` method in VField requires that both the `activeField` is null and the field block is the same `activeBlock` of the form. Therefore i used the `.leave` method to nullify the activeField attribut

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt [hunk 11]
```diff
@@ -932,4 +935,163 @@ class VBlockTests : VApplicationTestBase() {
...
+
+      singleBlockModel.abortTrail()
+      assertFalse(singleBlockModel.isRecordTrailed(0))
+      assertFalse(singleBlockModel.isRecordTrailed(singleBlockModel.currentRecord))
+    }
```
[the remark this answers is not in the document]
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt:1095]: Remove the redundancy of using the same record twice.

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/FormSampleTests.kt [hunk 12]
```diff
@@ -41,4 +71,161 @@ class FormSampleTests: JApplicationTestBase() {
...
+          if (it.isRecordTrailed(i))
+          {
+            return@all false
+          }
+        }
+        return@all true
+      })
+      SchemaUtils.drop(User)
+    }
+  }
```
[the remark this answers is not in the document]
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/form/FormSampleTests.kt:212]: Moved it to `FormWithMultipleBlockTest` and change the form to MultipleForm one

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt [hunk 13]
```diff
@@ -932,4 +935,163 @@ class VBlockTests : VApplicationTestBase() {
...
+
+  @Test
+  fun `setMode test`(){
+    formMultiple.model
+    val model = formMultiple.multipleBlock.vBlock
+
+    model.setMode(VConstants.MOD_INSERT)
+    assertEquals(model.getMode(), VConstants.MOD_INSERT)
+    assertTrue(model.fields.all {
+      for(counter in 1..model.recordCount) {
```
[the remark this answers is not in the document]
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt]: Done, add `onInsertHidden` to all fields in order to check if the fields access changes

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/examples/FormToTestSaveMultipleBlock.kt [hunk 14]
```diff
@@ -96,42 +96,49 @@ class FormToTestSaveMultipleBlock : DictionaryForm() {
...
       label = "ts"
       value = 0
       columns(c.ts)
+      onInsertHidden()
     }
     val uc = hidden(domain = INT(20)) {
       label = "uc"
       value = 0
       columns(c.uc)
+      onInsertHidden()
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/examples/FormToTestSaveMultipleBlock.kt:111]: remove OnInsertHidden() from Hidden fields

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/FormWithMultipleBlockTests.kt [hunk 15]
```diff
@@ -16,13 +16,65 @@
  */
 package org.kopi.galite.tests.form
 
+import org.jetbrains.exposed.sql.SchemaUtils
+import org.jetbrains.exposed.sql.insert
+import org.jetbrains.exposed.sql.transactions.transaction
+
 import kotlin.test.assertEquals
+import kotlin.test.assertTrue
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/form/FormWithMultipleBlockTests.kt:24]: reorder

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt [hunk 16]
```diff
@@ -1038,13 +1037,9 @@ class VBlockTests : VApplicationTestBase() {
...
-    transaction {
-      initSampleFormTables()
-      blockModel.enter()
-      assertEquals(blockModel, formSample.model.getActiveBlock())
-      assertEquals(blockModel.activeField, formSample.tb1.blockFields[3].vField)
-      SchemaUtils.drop(User)
-    }
+    blockModel.enter()
+    assertEquals(blockModel, formSample.model.getActiveBlock())
+    assertEquals(blockModel.activeField, formSample.tb1.blockFields[3].vField) // fields 1s -> 3rd are hidden
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VBlockTests.kt:1042]: replace : 
 `assertEquals(blockModel.activeField, formSample.tb1.name.vField)`

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt [hunk 17]
```diff
@@ -398,10 +398,12 @@ class VFieldTests : JApplicationTestBase() {
 
   @Test
   fun `enter test valid scenario` () {
-    val field = FormSample.tb1.blockFields[3].vField
 
+    val field = FormSample.tb1.blockFields[4].vField
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt:402]: replace : `FormSample.tb1.password.vField`

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt [hunk 18]
```diff
@@ -377,4 +381,57 @@ class VFieldTests : JApplicationTestBase() {
...
+  @Test
+  fun `fetchColumn existing column scenario test` () {
+    val field = FormSample.tb1.blockFields[0].vField
+
+    assertEquals(0, field.fetchColumn(User))
+  }
+
+  @Test
+  fun `fetchColumn not existing column scenario test` () {
+    val field = FormSample.tb1.blockFields[4].vField
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt:394]: replace by : `FormSample.tb1.password.vField`

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt [hunk 19]
```diff
@@ -377,4 +381,57 @@ class VFieldTests : JApplicationTestBase() {
 
     assertEquals(20, FormSample.tb4ToTestListDomain.age.value)
   }
+
+  @Test
+  fun `fetchColumn existing column scenario test` () {
+    val field = FormSample.tb1.blockFields[0].vField
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt:387]: replace by : replace : FormSample.tb1.id.vField

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt [hunk 20]
```diff
@@ -377,4 +381,57 @@ class VFieldTests : JApplicationTestBase() {
...
+    field.block?.activeField?.leave(false)
+    assertNull(field.block?.activeField)
+    field.enter()
+    assertEquals(field, field.block?.activeField)
+    assertTrue(field.hasFocus())
+  }
+
+  @Test
+  fun `enter test invalid scenario` () {
+    val field = FormSample.tb1.blockFields[3].vField
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt:414]: replace by : `FormSample.tb1.name.vField`

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt [hunk 21]
```diff
@@ -377,4 +381,57 @@ class VFieldTests : JApplicationTestBase() {
...
+
+    field.block?.enter()
+    assertFails { field.enter() }
+  }
+
+  @Test
+  fun `leave test valid scenario` () {
+    val field = FormSample.tb1.blockFields[3].vField // 3rd field in the first visible field
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt:422]: replace by : `FormSample.tb1.name.vField`

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt [hunk 22]
```diff
@@ -377,4 +381,57 @@ class VFieldTests : JApplicationTestBase() {
...
+
+    field.block?.enter()
+    field.leave(false)
+    assertNotEquals(field, field.block?.activeField)
+    assertNull(field.block?.activeField)
+  }
+
+  @Test
+  fun `leave test invalid scenario` () {
+    val field = FormSample.tb1.blockFields[4].vField
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt:432]: replace by : `FormSample.tb1.password.vField`

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt [hunk 23]
```diff
@@ -398,10 +398,12 @@ class VFieldTests : JApplicationTestBase() {
 
   @Test
   fun `enter test valid scenario` () {
-    val field = FormSample.tb1.blockFields[3].vField
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/form/VFieldTests.kt:401]: delete empty line

**Review by h-haddad — CHANGES_REQUESTED** [review]

**Review by hfazai — APPROVED** [review]

**Pull request merged**